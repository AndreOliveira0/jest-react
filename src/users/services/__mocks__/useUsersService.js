export const getUserById = jest.fn().mockResolvedValue({
  name: "Jos√©",
  username: "JO",
  email: "jose@teste.com",
});

export function useUsersService() {
  return { getUserById };
}