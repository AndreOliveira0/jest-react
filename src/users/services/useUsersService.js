import { api } from '../../app/clients';

function useUsersService() {

  const getUserById = id =>
    api.request({ url: `/users/${id}` })

  return { getUserById };
}

export { useUsersService };
