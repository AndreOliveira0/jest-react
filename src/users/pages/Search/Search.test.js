import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { getUserById } from '../../services/useUsersService';
import { Search } from '.';

jest.mock('../../services/useUsersService');

describe('component <Search />', () => {
  test('given render: match snapshot', () => {
    const { container } = render(<Search />);
    expect(container).toMatchSnapshot();
  });

  test('given a completed search: should display the message with the result', async () => {
    render(<Search />);
    const fieldSearch = screen.queryByTestId('search-field');
    const buttonFind = screen.queryByTestId('find-button');

    userEvent.type(fieldSearch, '1');
    userEvent.click(buttonFind);

    await waitFor(() => expect(getUserById).toHaveBeenCalled());
    expect(getUserById).toHaveBeenCalledWith('1');

    const successMessage = screen.queryByTestId('success-message');
    expect(successMessage).toBeInTheDocument();
    expect(successMessage.textContent).toBe(
      'O usuário encontrado foi: José',
    );
  });

  test('given an error in the search: should display an error message', async () => {
    console.error = jest.fn();
    const error = new Error('Falha ao buscar');
    getUserById.mockRejectedValueOnce(error);

    render(<Search />);
    const fieldSearch = screen.queryByTestId('search-field');
    const buttonFind = screen.queryByTestId('find-button');

    userEvent.type(fieldSearch, '2');
    userEvent.click(buttonFind);

    await waitFor(() => expect(getUserById).toHaveBeenCalled());
    expect(getUserById).toHaveBeenCalledWith('2');

    const errorMessage = screen.queryByTestId('error-message');
    expect(errorMessage).toBeInTheDocument();
    expect(errorMessage.textContent).toBe(
      'Erro ao buscar informações, tente novamente mais tarde.',
    );
    expect(console.error).toHaveBeenCalledWith(error);
    expect(buttonFind).toBeDisabled();
  });
});
