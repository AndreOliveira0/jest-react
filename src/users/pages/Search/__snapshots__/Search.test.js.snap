// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component <Search /> given render: match snapshot 1`] = `
<div>
  <main>
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm"
    >
      <div
        class="MuiBox-root MuiBox-root-1"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
        >
          Busca de usuário por ID:
        </p>
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            data-testid="search-field"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              placeholder="Digite aqui o ID do usuário"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
              style="padding-left: 8px;"
            >
              <legend
                class="PrivateNotchedOutline-legend-3"
                style="width: 0.01px;"
              >
                <span>
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-6"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
            data-testid="find-button"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Pesquisar
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root MuiBox-root-7"
        >
          <hr
            class="MuiDivider-root"
          />
        </div>
      </div>
    </div>
  </main>
</div>
`;
